README.rst
Scipion Grigoriefflab
=======================
This project is a python wrapper to use `Grigorieff Lab's <http://grigoriefflab.janelia.org/>`_ sofware
within Scipion framework.



CHANGES.txt
v1.0.0a0, 15/03/2018 -- First draft
v1.0.0, 8/5/2018 -- change plugin class
v1.0.0b2, 8/5/2018 -- small update
v1.0.0b3, 14/6/2018 -- get changes from scipion v 1.2.1
v1.0.0b4, 14/6/2018 -- name must be pip name
v1.0.0b5, 14/6/2018 -- test update

grigoriefflab/tests/test_protocols_grigoriefflab.py
# **************************************************************************
# *
# * Authors:    Laura del Cano (ldelcano@cnb.csic.es)
# *
# * Unidad de  Bioinformatica of Centro Nacional de Biotecnologia , CSIC
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 2 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************

from pyworkflow.em import *
from pyworkflow.tests import *
from pyworkflow.em.packages.grigoriefflab import *
from pyworkflow.em.protocol import ProtImportParticles, ProtImportVolumes


class TestBrandeisBase(BaseTest):
    @classmethod
    def setData(cls, dataProject='xmipp_tutorial'):
        cls.dataset = DataSet.getDataSet(dataProject)
        cls.micFn = cls.dataset.getFile('allMics')
        cls.volFn = cls.dataset.getFile('vol2')

    @classmethod
    def runImportMicrograph(cls, pattern, samplingRate, voltage,
                            scannedPixelSize, magnification,
                            sphericalAberration):
        """ Run an Import micrograph protocol. """
        # We have two options: pass the SamplingRate or
        # the ScannedPixelSize + microscope magnification
        kwargs = {
            'filesPath': pattern,
            'magnification': magnification,
            'voltage': voltage,
            'sphericalAberration': sphericalAberration
        }

        if samplingRate is not None:
            kwargs.update({'samplingRateMode': 0,
                           'samplingRate': samplingRate})
        else:
            kwargs.update({'samplingRateMode': 1,
                           'scannedPixelSize': scannedPixelSize})

        cls.protImport = ProtImportMicrographs(**kwargs)
        cls.launchProtocol(cls.protImport)

        # Check that input micrographs have been imported
        if cls.protImport.outputMicrographs is None:
            raise Exception('Import of micrograph: %s, failed. '
                            'outputMicrographs is None.' % pattern)

        return cls.protImport

    @classmethod
    def runImportVolumes(cls, pattern, samplingRate,
                         importFrom=ProtImportParticles.IMPORT_FROM_FILES):
        """ Run an Import particles protocol. """
        cls.protImport = cls.newProtocol(ProtImportVolumes,
                                         filesPath=pattern,
                                         samplingRate=samplingRate
                                         )
        cls.launchProtocol(cls.protImport)
        return cls.protImport

    @classmethod
    def runImportParticles(cls, pattern, samplingRate, checkStack=False,
                           importFrom=ProtImportParticles.IMPORT_FROM_FILES):
        """ Run an Import particles protocol. """
        if importFrom == ProtImportParticles.IMPORT_FROM_SCIPION:
            objLabel = 'from scipion (particles)'
        elif importFrom == ProtImportParticles.IMPORT_FROM_FILES:
            objLabel = 'from file (particles)'

        cls.protImport = cls.newProtocol(ProtImportParticles,
                                         objLabel=objLabel,
                                         filesPath=pattern,
                                         sqliteFile=pattern,
                                         samplingRate=samplingRate,
                                         checkStack=checkStack,
                                         importFrom=importFrom)

        cls.launchProtocol(cls.protImport)
        # Check that input images have been imported (a better way to do this?)
        if cls.protImport.outputParticles is None:
            raise Exception('Import of images: %s, failed. '
                            'outputParticles is None.' % pattern)
        return cls.protImport

    @classmethod
    def runImportMicrographBPV(cls, pattern):
        """ Run an Import micrograph protocol. """
        return cls.runImportMicrograph(pattern,
                                       samplingRate=1.237,
                                       voltage=300,
                                       sphericalAberration=2,
                                       scannedPixelSize=None,
                                       magnification=56000)

    @classmethod
    def runImportMicrographRCT(cls, pattern):
        """ Run an Import micrograph protocol. """
        return cls.runImportMicrograph(pattern,
                                       samplingRate=2.28,
                                       voltage=100,
                                       sphericalAberration=2.9,
                                       scannedPixelSize=None,
                                       magnification=50000)

    @classmethod
    def runImportParticleGrigorieff(cls, pattern):
        """ Run an Import micrograph protocol. """
        return cls.runImportParticles(pattern,
                                      samplingRate=4.,
                                      checkStack=True,
                                      importFrom=ProtImportParticles.IMPORT_FROM_SCIPION)

    @classmethod
    def runImportVolumesGrigorieff(cls, pattern):
        """ Run an Import micrograph protocol. """
        return cls.runImportVolumes(pattern,
                                    samplingRate=4.,
                                    importFrom=ProtImportParticles.IMPORT_FROM_FILES)


class TestImportParticles(BaseTest):
    @classmethod
    def setUpClass(cls):
        setupTestProject(cls)
        cls.dataset = DataSet.getDataSet('grigorieff')

    def test_import(self):
        parFile = self.dataset.getFile('particles/particles_iter_002.par')
        stackFile = self.dataset.getFile('particles/particles.mrc')

        protImport = self.newProtocol(ProtImportParticles,
                                      objLabel='import parfile & stk',
                                      parFile=parFile,
                                      stackFile=stackFile,
                                      samplingRate=9.90,
                                      importFrom=ProtImportParticles.IMPORT_FROM_FREALIGN)

        self.launchProtocol(protImport)
        # check that input images have been imported (a better way to do this?)
        outputParticles = getattr(protImport, 'outputParticles', None)

        if outputParticles is None:
            raise Exception('Import failed. Par file: %s' % parFile)

        self.assertTrue(outputParticles.getSize() == 180)

        goldFile = self.dataset.getFile('particles/particles.sqlite')
        goldSet = SetOfParticles(filename=goldFile)

        for p1, p2 in izip(goldSet,
                           outputParticles.iterItems(orderBy=['_micId', 'id'],
                                                     direction='ASC')):
            m1 = p1.getTransform().getMatrix()
            m2 = p2.getTransform().getMatrix()
            self.assertTrue(np.allclose(m1, m2, atol=0.01))

        self.assertTrue(outputParticles.hasCTF())
        self.assertTrue(outputParticles.hasAlignmentProj())


class TestBrandeisCtffind(TestBrandeisBase):
    @classmethod
    def setUpClass(cls):
        setupTestProject(cls)
        TestBrandeisBase.setData()
        cls.protImport = cls.runImportMicrographBPV(cls.micFn)

    def testCtffind(self):
        protCTF = ProtCTFFind(useCtffind4=False)
        protCTF.inputMicrographs.set(self.protImport.outputMicrographs)
        protCTF.ctfDownFactor.set(2)
        protCTF.numberOfThreads.set(4)
        self.proj.launchProtocol(protCTF, wait=True)
        self.assertIsNotNone(protCTF.outputCTF,
                             "SetOfCTF has not been produced.")

        valuesList = [[23861, 23664, 56],
                      [22383, 22153, 52.6],
                      [22716, 22526, 59.1]]
        for ctfModel, values in izip(protCTF.outputCTF, valuesList):
            self.assertAlmostEquals(ctfModel.getDefocusU(), values[0], delta=1000)
            self.assertAlmostEquals(ctfModel.getDefocusV(), values[1], delta=1000)
            self.assertAlmostEquals(ctfModel.getDefocusAngle(), values[2], delta=5)
            self.assertAlmostEquals(ctfModel.getMicrograph().getSamplingRate(),
                                    2.474, delta=0.001)

    def testCtffind2(self):
        protCTF = ProtCTFFind(useCtffind4=False)
        protCTF.inputMicrographs.set(self.protImport.outputMicrographs)
        protCTF.numberOfThreads.set(4)
        self.proj.launchProtocol(protCTF, wait=True)
        self.assertIsNotNone(protCTF.outputCTF, "SetOfCTF has not been produced.")

        valuesList = [[23920, 23499, 57.6], [22231, 21905, 59.3], [22444, 22270, 45.9]]
        for ctfModel, values in izip(protCTF.outputCTF, valuesList):
            self.assertAlmostEquals(ctfModel.getDefocusU(), values[0], delta=1000)
            self.assertAlmostEquals(ctfModel.getDefocusV(), values[1], delta=1000)
            self.assertAlmostEquals(ctfModel.getDefocusAngle(), values[2], delta=5)
            self.assertAlmostEquals(ctfModel.getMicrograph().getSamplingRate(),
                                    1.237, delta=0.001)


class TestBrandeisCtffind4(TestBrandeisBase):
    @classmethod
    def setUpClass(cls):
        setupTestProject(cls)
        TestBrandeisBase.setData()
        cls.protImport = cls.runImportMicrographBPV(cls.micFn)

    def testCtffind4V1(self):
        protCTF = ProtCTFFind()
        protCTF.inputMicrographs.set(self.protImport.outputMicrographs)
        protCTF.ctfDownFactor.set(2)
        protCTF.numberOfThreads.set(4)
        self.proj.launchProtocol(protCTF, wait=True)
        self.assertIsNotNone(protCTF.outputCTF, "SetOfCTF has not been produced.")

        valuesList = [[24049, 23586, 58], [22356, 22030, 44], [22661, 22480, 52]]
        for ctfModel, values in izip(protCTF.outputCTF, valuesList):
            self.assertAlmostEquals(ctfModel.getDefocusU(), values[0], delta=1000)
            self.assertAlmostEquals(ctfModel.getDefocusV(), values[1], delta=1000)
            self.assertAlmostEquals(ctfModel.getDefocusAngle(), values[2], delta=5)
            self.assertAlmostEquals(ctfModel.getMicrograph().getSamplingRate(),
                                    2.474, delta=0.001)

    def testCtffind4V22(self):
        protCTF = ProtCTFFind()
        protCTF.inputMicrographs.set(self.protImport.outputMicrographs)
        protCTF.numberOfThreads.set(4)
        self.proj.launchProtocol(protCTF, wait=True)
        self.assertIsNotNone(protCTF.outputCTF, "SetOfCTF has not been produced.")

        valuesList = [[23863, 23640, 54], [22159, 21983, 45.8], [22394, 22269, 171]]
        for ctfModel, values in izip(protCTF.outputCTF, valuesList):
            self.assertAlmostEquals(ctfModel.getDefocusU(), values[0], delta=1000)
            self.assertAlmostEquals(ctfModel.getDefocusV(), values[1], delta=1000)
            # def angle estimation differs a lot btw 4.0.x vs 4.1.x ctffind versions
            # self.assertAlmostEquals(ctfModel.getDefocusAngle(),values[2], delta=5)
            self.assertAlmostEquals(ctfModel.getMicrograph().getSamplingRate(),
                                    1.237, delta=0.001)


class TestBrandeisCtftilt(TestBrandeisBase):
    @classmethod
    def setUpClass(cls):
        setupTestProject(cls)
        cls.dataset = DataSet.getDataSet('rct')
        cls.micFn = cls.dataset.getFile('micrographs/F_rct_t_3003.tif')
        cls.protImport = cls.runImportMicrographRCT(cls.micFn)

    def testCtftilt(self):
        protCTF = self.newProtocol(ProtCTFTilt,
                                   minDefocus=1.0,
                                   maxDefocus=1.5,
                                   tiltA=38.0,
                                   ctfDownFactor=2,
                                   numberOfThreads=4)
        protCTF.inputMicrographs.set(self.protImport.outputMicrographs)
        self.proj.launchProtocol(protCTF, wait=True)
        self.assertIsNotNone(protCTF.outputCTF, "SetOfCTF has not been produced.")

        values = [14366, 12784, 21, 43.0]
        for ctfModel in protCTF.outputCTF:
            self.assertAlmostEquals(ctfModel.getDefocusU(), values[0], delta=1000)
            self.assertAlmostEquals(ctfModel.getDefocusV(), values[1], delta=1000)
            self.assertAlmostEquals(ctfModel.getDefocusAngle(), values[2], delta=5)
            self.assertAlmostEqual(float(ctfModel._ctftilt_tiltAngle), values[3], delta=3.0)
            self.assertAlmostEquals(ctfModel.getMicrograph().getSamplingRate(),
                                    4.56, delta=0.001)


class TestFrealignRefine(TestBrandeisBase):
    @classmethod
    def setUpClass(cls):
        setupTestProject(cls)
        dataProject = 'grigorieff'
        dataset = DataSet.getDataSet(dataProject)
        TestBrandeisBase.setData()
        particlesPattern = dataset.getFile('particles.sqlite')
        volFn = dataset.getFile('ref_volume.vol')
        cls.protImportPart = cls.runImportParticleGrigorieff(particlesPattern)
        cls.protImportVol = cls.runImportVolumesGrigorieff(volFn)

    def testFrealign(self):
        frealign = self.newProtocol(ProtFrealign,
                                    inputParticles=self.protImportPart.outputParticles,
                                    doInvert=False,
                                    input3DReference=self.protImportVol.outputVolume,
                                    useInitialAngles=True,
                                    mode=MOD_REFINEMENT,
                                    innerRadius=0.,
                                    outerRadius=241.,
                                    symmetry='C1',
                                    numberOfThreads=4,
                                    numberOfIterations=2,
                                    doWienerFilter=False,
                                    resolution=2.,
                                    highResolRefine=2.,
                                    resolClass=2.,
                                    writeMatchProjections=False,
                                    score=0,
                                    )
        frealign.inputParticles.set(self.protImportPart.outputParticles)
        frealign.input3DReference.set(self.protImportVol.outputVolume)
        self.launchProtocol(frealign)


class TestFrealignClassify(TestBrandeisBase):
    @classmethod
    def setUpClass(cls):
        setupTestProject(cls)
        dataProject = 'grigorieff'
        dataset = DataSet.getDataSet(dataProject)
        TestBrandeisBase.setData()
        particlesPattern = dataset.getFile('particles.sqlite')
        volFn = dataset.getFile('ref_volume.vol')
        cls.protImportPart = cls.runImportParticleGrigorieff(particlesPattern)
        cls.protImportVol = cls.runImportVolumesGrigorieff(volFn)

    def testFrealignClassify(self):
        frealign = self.newProtocol(ProtFrealignClassify,
                                    inputParticles=self.protImportPart.outputParticles,
                                    doInvert=False,
                                    input3DReference=self.protImportVol.outputVolume,
                                    numberOfIterations=3,
                                    itRefineAngles=2,
                                    itRefineShifts=3,
                                    numberOfClasses=2,
                                    useInitialAngles=True,
                                    mode=MOD_REFINEMENT,
                                    innerRadius=0.,
                                    outerRadius=241.,
                                    symmetry='C1',
                                    numberOfThreads=4,
                                    doWienerFilter=False,
                                    resolution=2.,
                                    highResolRefine=2.,
                                    resolClass=2.,
                                    writeMatchProjections=False,
                                    score=0,
                                    )
        frealign.inputParticles.set(self.protImportPart.outputParticles)
        frealign.input3DReference.set(self.protImportVol.outputVolume)
        self.launchProtocol(frealign)



grigoriefflab/plugin.py
from pyworkflow.plugin import Plugin
from bibtex import _bibtex  # Load bibtex dict with references

import os

MAGDIST_HOME_VAR = 'MAGDIST_HOME'
SUMMOVIE_HOME = 'SUMMOVIE_HOME'
UNBLUR_HOME = 'UNBLUR_HOME'
FREALIGN_HOME_VAR = 'FREALIGN_HOME'
CTFFIND4_HOME = 'CTFFIND4_HOME'
CTFFIND_HOME = 'CTFFIND_HOME'


class GrigoriefflabPlugin(Plugin):

    def __init__(self):
        configVars = {MAGDIST_HOME_VAR: os.path.join(os.environ['EM_ROOT'], 'mag_distortion-1.0.1'),
                      SUMMOVIE_HOME: os.path.join(os.environ['EM_ROOT'], 'summovie-1.0.2'),
                      UNBLUR_HOME: os.path.join(os.environ['EM_ROOT'], 'unblur-1.0.2'),
                      FREALIGN_HOME_VAR: os.path.join(os.environ['EM_ROOT'], 'frealign-9.07'),
                      CTFFIND4_HOME: os.path.join(os.environ['EM_ROOT'], 'ctffind4-4.1.10'),
                      CTFFIND_HOME: os.path.join(os.environ['EM_ROOT'], 'ctffind-3.6'),
                      }
        Plugin.__init__(self, 'scipion_grigoriefflab',
                              version="1.0.0b6",
                              logo="brandeis_logo.png",
                              configVars=configVars,
                              bibtex=_bibtex)

    def registerPluginBinaries(self, env):
        env.addPackage('ctffind', version='3.6',
                       tar='ctffind_V3.5.tgz')

        env.addPackage('ctffind4', version='4.0.15',
                       tar='ctffind_V4.0.15.tgz')

        env.addPackage('ctffind4', version='4.1.5',
                       tar='ctffind_V4.1.5.tgz')

        env.addPackage('ctffind4', version='4.1.8',
                       url="http://scipion.cnb.csic.es/downloads/scipion/software/em/ctffind_V4.1.8.tgz",
                       tar='ctffind_V4.1.8.tgz',
                       )

        env.addPackage('ctffind4', version='4.1.10',
                       default=True,  # Optional: install this binary by default
                       tar='ctffind4-4.1.10.tgz')

        env.addPackage('summovie', version='1.0.2',
                       tar='summovie_1.0.2.tgz')

        env.addPackage('unblur', version='1.0.15',
                       url="http://scipion.cnb.csic.es/downloads/scipion/software/em/unblur_1.0_150529.tgz",
                       tar='unblur_1.0_150529.tgz',
                       default=True  # Optional: install this binary by default
                       )

        env.addPackage('unblur', version='1.0.2',
                       tar='unblur_1.0.2.tgz')

        env.addPackage('frealign', version='9.07',
                       default=True,  # Optional: install this binary by default
                       tar='frealign_v9.07.tgz')

        env.addPackage('mag_distortion', version='1.0.1',
                       default=True,  # Optional: install this binary by default
                       tar='mag_distortion-1.0.1.tgz')


_plugin = GrigoriefflabPlugin()



grigoriefflab/__init__.py
# **************************************************************************
# *
# * Authors:     J.M. De la Rosa Trevin (jmdelarosa@cnb.csic.es)
# *
# * Unidad de  Bioinformatica of Centro Nacional de Biotecnologia , CSIC
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 2 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************
"""
This sub-package contains data and protocol classes
wrapping Grigorieff Lab programs at Brandeis
"""
from plugin import _plugin  # import at the very top
from grigoriefflab import *

from protocol_ctffind import ProtCTFFind
from protocol_ctftilt import ProtCTFTilt
from protocol_refinement import ProtFrealign
from protocol_magdist_estimate import ProtMagDistEst
from protocol_magdist_correct import ProtMagDistCorr
from protocol_magdist_correct_coords import ProtMagDistCorrCoord
from protocol_ml_classification import ProtFrealignClassify
from protocol_unblur import ProtUnblur
from protocol_summovie import ProtSummovie
from viewer import FrealignViewer, MagDistEstViewer, CtffindViewer
# Wizards
from wizard import *


grigoriefflab/grigoriefflab.py
# **************************************************************************
# *
# * Authors:     J.M. De la Rosa Trevin (jmdelarosa@cnb.csic.es)
# *
# * Unidad de  Bioinformatica of Centro Nacional de Biotecnologia , CSIC
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 2 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************

import os
from os.path import join, exists
from plugin import *  # Import everything from plugin.py - we'll need things that move there!


# Program common names
FREALIGN_APP = 'FREALIGN'
CTFFIND4_APP = 'CTFFIND4'
V4_0_15 = '4.0.15'
V4_1_10 = '4.1.10'


CTFFIND3 = 'ctffind3.exe'
CTFFIND3MP = 'ctffind3_mp.exe'
CTFFIND4 = 'ctffind'
CTFTILT = 'ctftilt.exe'
CTFTILTMP = 'ctftilt_mp.exe'
FREALIGN = 'frealign_v9.exe'
FREALIGNMP = 'frealign_v9_mp.exe'
MAGDISTEST = 'mag_distortion_estimate_openmp.exe'
MAGDISTCORR = 'mag_distortion_correct_openmp.exe'
CALC_OCC = 'calc_occ.exe'
RSAMPLE = 'rsample.exe'
UNBLUR = 'unblur_openmp.exe'
SUMMOVIE = 'sum_movie_openmp.exe'


def getVersion(var=FREALIGN_APP):
    varHome = var + '_HOME'
    path = os.environ[varHome]
    for v in getSupportedVersions(var):
        if v in path or v in os.path.realpath(path):
            return v
    return ''


def getSupportedVersions(var=FREALIGN_APP):
    if var == 'UNBLUR':
        return ['1.0_150529', '1.0.2']
    elif var == CTFFIND4_APP:
        return ['4.0.15', '4.1.5', '4.1.8', V4_1_10]
    else:  # FREALIGN
        return ['9.07']


def _getCtffind4():
    ctffind4 = join(os.environ[CTFFIND4_HOME], 'bin', CTFFIND4)
    if exists(ctffind4):
        return ctffind4
    else:
        return join(os.environ[CTFFIND4_HOME], CTFFIND4)


def _getHome(key, default):
    """ Get the required home path, if not present..
    the default value will be used from EM_ROOT.
    """
    return os.environ.get(key, join(os.environ['EM_ROOT'], default))

CTFFIND_PATH = join(os.environ[CTFFIND_HOME], CTFFIND3)
CTFFINDMP_PATH = join(os.environ[CTFFIND_HOME], CTFFIND3MP)
CTFFIND4_PATH = _getCtffind4()

CTFTILT_PATH = join(os.environ[CTFFIND_HOME], CTFTILT)
CTFTILTMP_PATH = join(os.environ[CTFFIND_HOME], CTFTILTMP)

FREALIGN_HOME = _getHome(FREALIGN_HOME_VAR, 'frealign')
FREALIGN_PATH = join(FREALIGN_HOME, 'bin', FREALIGN)
FREALIGNMP_PATH = join(FREALIGN_HOME, 'bin', FREALIGNMP)
CALC_OCC_PATH = join(FREALIGN_HOME, 'bin', CALC_OCC)
RSAMPLE_PATH = join(FREALIGN_HOME, 'bin', RSAMPLE)

MAGDIST_HOME = _getHome(MAGDIST_HOME_VAR, 'mag_distortion')
MAGDISTEST_PATH = join(MAGDIST_HOME, 'bin', MAGDISTEST)
MAGDISTCORR_PATH = join(MAGDIST_HOME, 'bin', MAGDISTCORR)

UNBLUR_PATH = join(_getHome(UNBLUR_HOME, 'unblur'), 'bin', UNBLUR)
SUMMOVIE_PATH = join(_getHome(SUMMOVIE_HOME, 'summovie'), 'bin', SUMMOVIE)


def validateMagDistorsionInstallation():
    """ Check if the installation of this protocol is correct.
    Can't rely on package function since this is a "multi package" package
    Returning an empty list means that the installation is correct
    and there are not errors. If some errors are found, a list with
    the error messages will be returned.
    """
    missingPaths = []

    if not os.path.exists(MAGDIST_HOME):
        missingPaths.append("%s : %s" % (MAGDIST_HOME_VAR, MAGDIST_HOME))
    return missingPaths


grigoriefflab/wizard.py
# **************************************************************************
# *
# * Authors:     Jose Gutierrez (jose.gutierrez@cnb.csic.es)
# *
# * Unidad de  Bioinformatica of Centro Nacional de Biotecnologia , CSIC
# *
# * This program is free software; you can redistribute it and/or modify
# * it under the terms of the GNU General Public License as published by
# * the Free Software Foundation; either version 2 of the License, or
# * (at your option) any later version.
# *
# * This program is distributed in the hope that it will be useful,
# * but WITHOUT ANY WARRANTY; without even the implied warranty of
# * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# * GNU General Public License for more details.
# *
# * You should have received a copy of the GNU General Public License
# * along with this program; if not, write to the Free Software
# * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
# * 02111-1307  USA
# *
# *  All comments concerning this program package may be sent to the
# *  e-mail address 'scipion@cnb.csic.es'
# *
# **************************************************************************
"""
This module implement some wizards
"""

import os
import Tkinter as tk
import ttk

from pyworkflow.em.constants import *
from constants import *

from protocol_ctffind import ProtCTFFind
from protocol_ctftilt import ProtCTFTilt
import pyworkflow.gui.dialog as dialog
from pyworkflow.em.wizard import *
from protocol_refinement import ProtFrealign
from protocol_ml_classification import ProtFrealignClassify
from protocol_magdist_estimate import ProtMagDistEst
from protocol_magdist_correct import ProtMagDistCorr

from pyworkflow import findResource


# ===============================================================================
# CTFs
# ===============================================================================

class BrandeisCTFWizard(CtfWizard):
    _targets = [(ProtCTFFind, ['ctfDownFactor', 'lowRes', 'highRes']),
                (ProtCTFTilt, ['ctfDownFactor', 'lowRes', 'highRes'])]

    def _getParameters(self, protocol):

        label, value = self._getInputProtocol(self._targets, protocol)

        protParams = {}
        protParams['input'] = protocol.inputMicrographs
        protParams['label'] = label
        protParams['value'] = value
        return protParams

    def _getProvider(self, protocol):
        _objs = self._getParameters(protocol)['input']
        return CtfWizard._getListProvider(self, _objs)

    def show(self, form):
        protocol = form.protocol
        params = self._getParameters(protocol)
        _value = params['value']
        _label = params['label']

        #        form.setParamFromVar('inputMicrographs') # update selected input micrographs
        provider = self._getProvider(protocol)

        if provider is not None:
            args = {'unit': UNIT_PIXEL,
                    'downsample': _value[0],
                    'lf': _value[1],
                    'hf': _value[2],
                    'showInAngstroms': True
                    }
            d = CtfDownsampleDialog(form.root, provider, **args)

            if d.resultYes():
                form.setVar(_label[0], d.getDownsample())
                form.setVar(_label[1], d.getLowFreq())
                form.setVar(_label[2], d.getHighFreq())
        else:
            dialog.showWarning("Empty input", "Select elements first", form.root)

    @classmethod
    def getView(self):
        return "wiz_ctf_downsampling"

    # ===============================================================================


# MASKS
# ===============================================================================

class FrealignVolRadiiWizard(VolumeMaskRadiiWizard):
    _targets = [(ProtFrealign, ['innerRadius', 'outerRadius']),
                (ProtFrealignClassify, ['innerRadius', 'outerRadius'])]

    def _getParameters(self, protocol):
        label, value = self._getInputProtocol(self._targets, protocol)

        protParams = {}
        protParams['input'] = protocol.input3DReference
        protParams['label'] = label
        protParams['value'] = value
        return protParams

    def _getProvider(self, protocol):
        _objs = self._getParameters(protocol)['input']
        return VolumeMaskRadiiWizard._getListProvider(self, _objs)

    def show(self, form):
        params = self._getParameters(form.protocol)
        _value = params['value']
        _label = params['label']
        VolumeMaskRadiiWizard.show(self, form, _value, _label, UNIT_ANGSTROM)


# ===============================================================================
# FILTERS
# ===============================================================================

class FrealignBandpassWizard(FilterParticlesWizard):
    _targets = [(ProtFrealign, ['lowResolRefine', 'highResolRefine']),
                (ProtFrealignClassify, ['lowResolRefine', 'highResolRefine'])]

    def _getParameters(self, protocol):

        label, value = self._getInputProtocol(self._targets, protocol)

        protParams = {}
        protParams['input'] = protocol.inputParticles
        protParams['label'] = label
        protParams['value'] = value
        protParams['mode'] = FILTER_NO_DECAY
        return protParams

    def _getProvider(self, protocol):
        _objs = self._getParameters(protocol)['input']
        return FilterParticlesWizard._getListProvider(self, _objs)

    def show(self, form):
        protocol = form.protocol
        provider = self._getProvider(protocol)
        params = self._getParameters(protocol)

        if provider is not None:

            args = {'mode': params['mode'],
                    'lowFreq': params['value'][0],
                    'highFreq': params['value'][1],
                    'unit': UNIT_ANGSTROM
                    }

            args['showDecay'] = False

            d = BandPassFilterDialog(form.root, provider, **args)

            if d.resultYes():
                form.setVar('highResolRefine', d.samplingRate / d.getHighFreq())
                form.setVar('lowResolRefine', d.samplingRate / d.getLowFreq())

        else:
            dialog.showWarning("Input particles", "Select particles first", form.root)


class FrealignVolBandpassWizard(FilterVolumesWizard):
    _targets = [(ProtFrealign, ['resolution']),
                (ProtFrealignClassify, ['resolution'])]

    def _getParameters(self, protocol):

        label, value = self._getInputProtocol(self._targets, protocol)

        protParams = {}
        protParams['input'] = protocol.input3DReference
        protParams['label'] = label
        protParams['value'] = value
        protParams['mode'] = FILTER_LOW_PASS_NO_DECAY
        return protParams

    def _getProvider(self, protocol):
        _objs = self._getParameters(protocol)['input']
        return FilterVolumesWizard._getListProvider(self, _objs)

    def show(self, form):
        params = self._getParameters(form.protocol)
        protocol = form.protocol
        provider = self._getProvider(protocol)

        if provider is not None:

            args = {'mode': params['mode'],
                    'highFreq': params['value'],
                    'unit': UNIT_ANGSTROM
                    }

            args['showLowFreq'] = False
            args['showDecay'] = False

            d = BandPassFilterDialog(form.root, provider, **args)

            if d.resultYes():
                form.setVar('resolution', d.samplingRate / d.getHighFreq())

        else:
            dialog.showWarning("Input volumes", "Select volumes first", form.root)


# ===============================================================================
# Gold diffraction ring
# ===============================================================================

class BrandeisDistWizard(CtfWizard):
    _targets = [(ProtMagDistEst, ['lowRes', 'highRes'])]

    def _getParameters(self, protocol):

        label, value = self._getInputProtocol(self._targets, protocol)

        protParams = {}
        protParams['input'] = protocol.inputMicrographs
        protParams['label'] = label
        protParams['value'] = value
        protParams['pix'] = protocol.inputMicrographs.get().getSamplingRate()
        return protParams

    def _getProvider(self, protocol):
        _objs = self._getParameters(protocol)['input']
        return CtfWizard._getListProvider(self, _objs)

    def show(self, form):
        protocol = form.protocol
        params = self._getParameters(protocol)
        _value = params['value']
        _label = params['label']
        pix = params['pix']

        provider = self._getProvider(protocol)

        if provider is not None:
            args = {'unit': UNIT_ANGSTROM,
                    'lf': "%0.2f" % (pix / _value[0]),
                    'hf': "%0.2f" % (pix / _value[1])
                    }
            d = CtfDialog(form.root, provider, **args)

            if d.resultYes():
                form.setVar(_label[0], "%0.2f" % (pix / d.getLowFreq()))
                form.setVar(_label[1], "%0.2f" % (pix / d.getHighFreq()))
        else:
            dialog.showWarning("Empty input", "Select elements first", form.root)

    @classmethod
    def getView(self):
        return "wiz_ctf"


# ===============================================================================
# Pixel size
# ===============================================================================

class BrandeisPixSizeWizard(Wizard):
    _targets = [(ProtMagDistCorr, ['newPix'])]

    def show(self, form):
        form.setVar('newPix', form.protocol.inputMovies.get().getSamplingRate())
